crates/bevy_core_pipeline/src/tonemapping/node.rs

```diff
@@ -0,0 +1,153 @@
+use std::sync::Mutex;
+
+use bevy_ecs::prelude::*;
+use bevy_ecs::query::QueryState;
+use bevy_render::{
+    render_graph::{Node, NodeRunError, RenderGraphContext, SlotInfo, SlotType, SlotValue},
+    render_resource::{
+        BindGroup, BindGroupDescriptor, BindGroupEntry, BindingResource, LoadOp, Operations,
+        RenderPassColorAttachment, RenderPassDescriptor, RenderPipelineCache, SamplerDescriptor,
+        TextureViewId,
+    },
+    renderer::RenderContext,
+    view::{ExtractedView, ViewMainTexture, ViewTarget},
+};
+
+use super::{TonemappingPipeline, TonemappingTarget};
+
+pub struct TonemappingNode {
+    query: QueryState<(&'static ViewTarget, &'static TonemappingTarget), With<ExtractedView>>,
+    cached_texture_bind_group: Mutex<Option<(TextureViewId, BindGroup)>>,
+}
+
+impl TonemappingNode {
+    pub const IN_VIEW: &'static str = "view";
+    pub const IN_TEXTURE: &'static str = "in_texture";
+    pub const OUT_TEXTURE: &'static str = "out_texture";
+
+    pub fn new(world: &mut World) -> Self {
+        Self {
+            query: QueryState::new(world),
+            cached_texture_bind_group: Mutex::new(None),
+        }
+    }
+}
+
+impl Node for TonemappingNode {
+    fn input(&self) -> Vec<SlotInfo> {
+        vec![
+            SlotInfo::new(TonemappingNode::IN_TEXTURE, SlotType::TextureView),
+            SlotInfo::new(TonemappingNode::IN_VIEW, SlotType::Entity),
+        ]
+    }
+
+    fn output(&self) -> Vec<SlotInfo> {
+        vec![SlotInfo::new(
+            TonemappingNode::OUT_TEXTURE,
+            SlotType::TextureView,
+        )]
+    }
+
+    fn update(&mut self, world: &mut World) {
+        self.query.update_archetypes(world);
+    }
+
+    fn run(
+        &self,
+        graph: &mut RenderGraphContext,
+        render_context: &mut RenderContext,
+        world: &World,
+    ) -> Result<(), NodeRunError> {
+        let view_entity = graph.get_input_entity(Self::IN_VIEW)?;
+        let in_texture = graph.get_input_texture(Self::IN_TEXTURE)?;
+
+        let render_pipeline_cache = world.get_resource::<RenderPipelineCache>().unwrap();
+        let tonemapping_pipeline = world.get_resource::<TonemappingPipeline>().unwrap();
+
+        let (target, tonemapping_target) = match self.query.get_manual(world, view_entity) {
+            Ok(query) => query,
+            Err(_) => return Ok(()),
+        };
+
+        let pipeline = match render_pipeline_cache.get(tonemapping_target.pipeline) {
+            Some(pipeline) => pipeline,
+            None => return Ok(()),
+        };
```

Comment:
For the query I was copying the `MainPass3dNode` which does
```rust
        let (opaque_phase, alpha_mask_phase, transparent_phase, target, depth) =
            match self.query.get_manual(world, view_entity) {
                Ok(query) => query,
                Err(_) => return Ok(()), // No window
            };
```

The pipeline could not be ready yet because it is still resolving imports or something.